actionview/lib/action_view/helpers/tag_helper.rb

```diff
@@ -46,8 +49,168 @@ class TagBuilder # :nodoc:
         include CaptureHelper
         include OutputSafetyHelper
 
-        HTML_VOID_ELEMENTS = %i(area base br col embed hr img input keygen link meta param source track wbr).to_set
-        SVG_SELF_CLOSING_ELEMENTS = %i(animate animateMotion animateTransform circle ellipse line path polygon polyline rect set stop use view).to_set
+        def self.define_element(name, code_generator:, method_name: name.to_s.underscore)
+          code_generator.define_cached_method(method_name, namespace: :tag_builder) do |batch|
+            batch.push(<<~RUBY) unless instance_methods.include?(method_name.to_sym)
+              def #{method_name}(content = nil, escape: true, **options, &block)
+                tag_string(#{name.inspect}, content, escape: escape, **options, &block)
+              end
+            RUBY
+          end
+        end
+
+        def self.define_void_element(name, code_generator:, method_name: name.to_s.underscore, self_closing: false)
+          code_generator.define_cached_method(method_name, namespace: :tag_builder) do |batch|
+            batch.push(<<~RUBY)
+              def #{method_name}(content = nil, escape: true, **options, &block)
+                if content.present? || block
+                  tag_string(#{name.inspect}, content, escape: escape, **options, &block)
+                else
+                  void_tag_string(#{name.inspect}, options, escape, #{self_closing})
+                end
+              end
+            RUBY
+          end
+        end
+
+        def self.define_self_closing_element(name, **options)
+          define_void_element(name, self_closing: true, **options)
+        end
+
+        ActiveSupport::CodeGenerator.batch(self, __FILE__, __LINE__) do |code_generator|
+          with_options(code_generator: code_generator) do
```

Comment:
Is the `with_options` really worth it here?

Reply:
I wasn't a fan of copy-pasting the `code_generator:` option across all the calls, and didn't think there was a cost to depending on an Active Support nicety internally. I'll unwrap this block if we think it's necessary.

Reply:
Removed.